pkg/project/auth/cache.go

```diff
@@ -0,0 +1,359 @@
+	reviewRecordStore       cache.Store
+	userSubjectRecordStore  cache.Store
+	groupSubjectRecordStore cache.Store
+
+	masterNamespace              string
+	masterBindingResourceVersion string
+	masterPolicyResourceVersion  string
+
+	reviewer Reviewer
+
+	syncHandler func(request *reviewRequest, userSubjectRecordStore cache.Store, groupSubjectRecordStore cache.Store, reviewRecordStore cache.Store) error
+}
+
+// reviewRecordKeyFunc is a key func for reviewRecord objects
+func reviewRecordKeyFunc(obj interface{}) (string, error) {
+	reviewRecord, ok := obj.(*reviewRecord)
+	if !ok {
+		return "", fmt.Errorf("expected reviewRecord")
+	}
+	return reviewRecord.namespace, nil
+}
+
+// subjectRecordKeyFunc is a key func for subjectRecord objects
+func subjectRecordKeyFunc(obj interface{}) (string, error) {
+	subjectRecord, ok := obj.(*subjectRecord)
+	if !ok {
+		return "", fmt.Errorf("expected subjectRecord")
+	}
+	return subjectRecord.subject, nil
+}
+
+// NewAuthorizationCache creates a new AuthorizationCache
+// TODO: policyStore should be an IndexerByNamespace after rebase
+// TODO: bindingStore should be an IndexerByNamespace after rebase
+func NewAuthorizationCache(reviewer Reviewer, namespaceStore cache.Store, bindingStore cache.Store, policyStore cache.Store, masterNamespace string) *AuthorizationCache {
+	result := &AuthorizationCache{
+		namespaceStore:               namespaceStore,
+		bindingStore:                 bindingStore,
+		policyStore:                  policyStore,
+		reviewRecordStore:            cache.NewStore(reviewRecordKeyFunc),
+		userSubjectRecordStore:       cache.NewStore(subjectRecordKeyFunc),
+		groupSubjectRecordStore:      cache.NewStore(subjectRecordKeyFunc),
+		masterNamespace:              masterNamespace,
+		masterBindingResourceVersion: "",
+		masterPolicyResourceVersion:  "",
+	}
+	result.syncHandler = result.syncRequest
+	return result
+}
+
+// Run begins watching and synchronizing the cache
+func (ac *AuthorizationCache) Run(period time.Duration) {
+	go util.Forever(func() { ac.synchronize() }, period)
+}
+
+// synchronize runs a a full synchronization over the cache data.  it must be run in a single-writer model, it's not thread-safe by design.
+func (ac *AuthorizationCache) synchronize() {
+	// TODO: when caches support a high-water mark, do not run sync if no cache's highwater mark has changed
+	userSubjectRecordStore := ac.userSubjectRecordStore
+	groupSubjectRecordStore := ac.groupSubjectRecordStore
+	reviewRecordStore := ac.reviewRecordStore
+
+	// TODO: when policyStore and bindingStore are indexers, check if masterResourceVersions have changed to know to invalidate the entire cache
+	invalidateCache := false
+	if invalidateCache {
+		userSubjectRecordStore = cache.NewStore(subjectRecordKeyFunc)
+		groupSubjectRecordStore = cache.NewStore(subjectRecordKeyFunc)
+		reviewRecordStore = cache.NewStore(reviewRecordKeyFunc)
+	}
+
+	namespaceSet := util.NewStringSet()
+	items := ac.namespaceStore.List()
```

Comment:
Either scope items in the for loop or name it more precisely.  The reuse makes it harder to read.